hibernate-core/src/test/java/org/hibernate/orm/test/join/JoinWithSingleTableInheritanceTest.java

```diff
@@ -0,0 +1,256 @@
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+/**
+ * @author Jan Schatteman
+ */
+@TestForIssue( jiraKey = "HHH-16435")
+@DomainModel(
+		annotatedClasses = {
+				JoinWithSingleTableInheritanceTest.AbstractSuperClass.class,
+				JoinWithSingleTableInheritanceTest.ChildEntityA.class,
+				JoinWithSingleTableInheritanceTest.ChildEntityB.class,
+				JoinWithSingleTableInheritanceTest.RootOne.class
+		}
+)
+@SessionFactory
+public class JoinWithSingleTableInheritanceTest {
+
+	@AfterEach
+	public void cleanup( SessionFactoryScope scope ) {
+		scope.inTransaction(
+				s -> {
+					s.createMutationQuery( "delete from RootOne" ).executeUpdate();
+					s.createMutationQuery( "delete from AbstractSuperClass" ).executeUpdate();
+				}
+		);
+	}
+
+	@Test
+	public void testLeftJoinOnSingleTableInheritance(SessionFactoryScope scope) {
+		scope.inTransaction(
+				s -> s.persist( new RootOne(1) )
+		);
+
+		scope.inTransaction(
+				s -> {
+					List<RootOne> l = s.createSelectionQuery( "select r from RootOne r left join ChildEntityA ce on ce.id = r.someOtherId", RootOne.class ).list();
+					assertEquals( 1, l.size() );
+				}
+		);
+	}
+
+	@Test
+	public void testLeftJoinOnSingleTableInheritance2(SessionFactoryScope scope) {
+		scope.inTransaction(
+				s -> {
+					s.persist(new ChildEntityA( 11 ));
+					s.persist(new ChildEntityB( 21 ));
+					RootOne r1 = new RootOne(1);
+					r1.setSomeOtherId( 11 );
+					s.persist( r1 );
+					RootOne r2 = new RootOne(2);
+					r2.setSomeOtherId( 21 );
+					s.persist( r2 );
+				}
+		);
+
+		scope.inTransaction(
+				s -> {
+					List<Object[]> l = s.createSelectionQuery( "select r.id, r.someOtherId, ce.id, ce.disc_col from RootOne r left join ChildEntityA ce on ce.id = r.someOtherId order by r.id", Object[].class ).list();
+					assertEquals( 2, l.size() );
+					Object[] r1 = l.get( 0 );
+					assertEquals( 1, (int) r1[0] );
+					assertEquals( 11, (int) r1[1] );
+					assertEquals( 11, (int) r1[2] );
+					assertEquals( 1, (int) r1[3] );
+					// r2 has to be there, but shouldn't have any data w/ respect to ChildEntityB
+					Object[] r2 = l.get( 1 );
+					assertEquals( 2, (int) r2[0] );
+					assertEquals( 21, (int) r2[1] );
+					assertNull( r2[2] );
+					assertNull( r2[3] );
+				}
+		);
+	}
+
+	@Test
+	public void testCrossJoinOnSingleTableInheritance(SessionFactoryScope scope) {
+		scope.inTransaction(
+				s -> {
+					s.persist(new ChildEntityA( 11 ));
+					s.persist(new ChildEntityA( 12 ));
+					s.persist(new ChildEntityB( 21 ));
+					RootOne r1 = new RootOne(1);
+					r1.setSomeOtherId( 11 );
+					s.persist( r1 );
+					RootOne r2 = new RootOne(2);
+					r2.setSomeOtherId( 12 );
+					s.persist( r2 );
+					RootOne r3 = new RootOne(3);
+					r3.setSomeOtherId( 21 );
+					s.persist( r3 );
+				}
+		);
+
+		scope.inTransaction(
+				s -> {
+					List<Object[]> l = s.createSelectionQuery( "select r.id, r.someOtherId from RootOne r join ChildEntityA", Object[].class ).list();
```

Comment:
A cross join needs to be spelled out i.e. `RootOne r cross join ChildEntityA a`. Usually, only a cross join may omit the `on` clause, so this should IMO actually fail with a `SemanticQueryException`, thought that's a different topic for which you can start a discussion and create a JIRA issue. In this test, you don't need to test cross joins. I figured that `consumeEntityJoin` is never called with `SqlAstJoinType.CROSS`

Reply:
Indeed I saw that for those tests it seemed to always be 'inner'

Reply:
A query with an entity join that is lacking an `on` clause, like your example, should fail with a `SemanticQueryException`, because it is only legal to omit the `on` clause for cross joins.